org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/BuildNotifier.java

```diff
@@ -76,8 +89,35 @@ public void begin() {
  * Check whether the build has been canceled.
  */
 public void checkCancel() {
-	if (this.monitor != null && this.monitor.isCanceled())
-		throw new OperationCanceledException();
+	if (this.monitor != null && this.monitor.isCanceled()) {
+		// User requested cancel.
+		if (this.buildKind == IncrementalProjectBuilder.AUTO_BUILD) {
+			// Since jdt can not resume the incremental build we follow that request only after a period of grace
+			// to increase the chance the cancel happens between project builds and not during a project build
+			if (getBuildDurationInMs() > millisecondsBeforeCancelAutoBuild) {
+				throw new OperationCanceledException();
+			}
+		} else {
+			// non autobuild will be canceled immediately.
+			throw new OperationCanceledException();
+		}
+	}
+	if (this.interruptSupplier.getAsBoolean()) {
+		// Automatic request to interrupt build due to a conflicting user action.
+		// Since jdt can not resume the incremental build we follow that request only after a period of grace:
+		if (getBuildDurationInMs() > millisecondsBeforeInterruptAutoBuild) {
+			// This is a trade off between UI responsiveness and additional work to restart autobuild.
+			// We cancel autobuild to avoid possible UI freeze where the user can not manually cancel.
+			throw new OperationCanceledException();
+		}
+		// XXX we could also use this.workDone to get an estimate how long the build will need and cancel immediately if
+		// that duration is greater then the period of grace.
+		// But since that estimate may not be very accurate it may be better to just wait the period of grace.
+	}
+}
+
+private int getBuildDurationInMs() {
+	return (int) ((System.nanoTime() - this.startTimeNanos) / 1_000_000);
```

Comment:
The cast to in seems unneccessary.

Reply:
Type mismatch: cannot convert from long to int

Reply:
Why not just let the method return long instead of int?

Reply:
done. i wonder why it works to compare int with long without typecast